src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs

```diff
@@ -5512,6 +5538,86 @@ protected override void Cleanup()
         }
         #endregion
 
+        #region WithCancellation
+        /// <summary>Task that's completed when another task, timeout, or cancellation token triggers.</summary>
+        private protected sealed class CancellationPromise<TResult> : Task<TResult>, ITaskCompletionAction
+        {
+            /// <summary>The source task.  It's stored so that we can remove the continuation from it upon timeout or cancellation.</summary>
+            private readonly Task _task;
+            /// <summary>Cancellation registration used to unregister from the token source upon timeout or the task completing.</summary>
+            private readonly CancellationTokenRegistration _registration;
+            /// <summary>The timer used to implement the timeout.  It's stored so that it's rooted and so that we can dispose it upon cancellation or the task completing.</summary>
+            private readonly TimerQueueTimer? _timer;
+
+            internal CancellationPromise(Task source, uint millisecondsDelay, CancellationToken token)
+            {
+                Debug.Assert(source != null);
+                Debug.Assert(millisecondsDelay != 0);
+
+                // Register with the target task.
+                _task = source;
+                source.AddCompletionAction(this);
+
+                // Register with a timer if it's needed.
+                if (millisecondsDelay != Timeout.UnsignedInfinite)
+                {
+                    _timer = new TimerQueueTimer(static state =>
+                    {
+                        var thisRef = (CancellationPromise<TResult>)state!;
+                        if (thisRef.TrySetException(new TimeoutException()))
+                        {
+                            thisRef.Cleanup();
+                        }
+                    }, this, millisecondsDelay, Timeout.UnsignedInfinite, flowExecutionContext: false);
+                }
+
+                // Register with the cancellation token.
+                _registration = token.UnsafeRegister(static (state, cancellationToken) =>
```

Comment:
Random C# question: Do you recommending marking lambdas that don't need state as static? I think this kind of method definitely benefits from it, but I wonder about `var names = people.Select(p => p.Name)`

Reply:
I personally like to use static now anywhere it's not 100% obvious that no state is being captured, which generally means if the body is anything more than an expression I prefer to use static.